### PR TITLE
Add CBMC harness for QueueGiveFromISR

### PR DESCRIPTION
**Pull request by mmuesly**
Description
-----------

Add a CBMC harness for QueueGiveFromISR. See the added README.md for assumptions.

Depends on #774, #697, #688.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/cbmc/include/queue_init.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "queue_datastructure.h"
+
+/* Using prvCopyDataToQueue together with prvNotifyQueueSetContainer
+   leads to a problem space explosion. Therefore, we use this stub
+   and a sepearted proof on prvCopyDataToQueue to deal with it.
+   As prvNotifyQueueSetContainer is disabled if configUSE_QUEUE_SETS != 1,
+   in other cases the original implementation should be used. */
+#if( configUSE_QUEUE_SETS == 1 )
```
Comment (danielsn) [inline tools/cbmc/include/queue_init.h]: I assume the real version changes some set of bytes.  I wonder if you could get most of the effect by havocing a small fix number of randomly selected bytes.  If not, there should be some explanation for why this is sound even tho you're not havocing anything.

File: tools/cbmc/include/queue_init.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	BaseType_t prvCopyDataToQueue( Queue_t * const pxQueue, const void *pvItemToQueue, const BaseType_t xPosition )
+	{
+		if(pxQueue->uxItemSize > ( UBaseType_t ) 0)
+		{
+			__CPROVER_assert(__CPROVER_r_ok(pvItemToQueue, ( size_t ) pxQueue->uxItemSize), "pvItemToQueue region must be readable");
+			if(xPosition == queueSEND_TO_BACK){
+				__CPROVER_assert(__CPROVER_w_ok(( void * ) pxQueue->pcWriteTo, ( size_t ) pxQueue->uxItemSize), "pxQueue->pcWriteTo region must be writable");
```
Comment (danielsn) [inline tools/cbmc/include/queue_init.h]: These lines are too long to read. Wrap them. Is there no auto-formatter for FreeRTOS?

File: tools/cbmc/include/queue_init.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			}else{
+				__CPROVER_assert(__CPROVER_w_ok(( void * ) pxQueue->u.xQueue.pcReadFrom, ( size_t ) pxQueue->uxItemSize), "pxQueue->u.xQueue.pcReadFrom region must be writable");
```
Comment (danielsn) [inline tools/cbmc/include/queue_init.h]: Is the cast to void* necessary?

File: tools/cbmc/include/queue_init.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	uint8_t ucQueueType;
+	__CPROVER_assume(uxQueueLength > 0);
+	__CPROVER_assume(uxItemSize < uxItemSizeBound);
+	/* The implicit assumption for the QueueGenericCreate method is,
+	   that there are no overflows in the computation and the inputs are safe.
+	   There is no check for this in the code base */
+	UBaseType_t upper_bound = portMAX_DELAY - sizeof(Queue_t);
+	__CPROVER_assume(uxItemSize < (upper_bound)/uxQueueLength);
+	QueueHandle_t xQueue =
+		xQueueGenericCreate(uxQueueLength, uxItemSize, ucQueueType);
```
Comment (danielsn) [inline tools/cbmc/include/queue_init.h]: Should this be unconstrained?

File: tools/cbmc/include/queue_init.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		#endif
+	}
+	return xQueue;
+}
+
+/* Create a mostly unconstrained Mutex */
+QueueHandle_t xUnconstrainedMutex( void ) {
+	uint8_t ucQueueType;
+	QueueHandle_t xQueue =
+		xQueueCreateMutex(ucQueueType);
```
Comment (danielsn) [inline tools/cbmc/include/queue_init.h]: should this be unconstrained.

File: tools/cbmc/proofs/CBMCStubLibrary/tasksStubs.c [hunk 6]
```diff
@@ -0,0 +1,47 @@
+#include "FreeRTOS.h"
+#include "task.h"
+#include "tasksStubs.h"
+
+#ifndef TASK_STUB_COUNTER
+	#define TASK_STUB_COUNTER 0;
```
Comment (danielsn) [inline tools/cbmc/proofs/CBMCStubLibrary/tasksStubs.c]: Why is this 0? What would it mean if it were non-zero? Add a comment.

File: tools/cbmc/proofs/Queue/QueueGiveFromISR/QueueGiveFromISR_harness.c [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "queue_init.h"
+
+#include "cbmc.h"
+
+void harness(){
+	QueueHandle_t xQueue = xUnconstrainedMutex();
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGiveFromISR/QueueGiveFromISR_harness.c]: Can these be ```null```?

File: tools/cbmc/proofs/Queue/QueueGiveFromISR/QueueGiveFromISR_harness.c [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+	BaseType_t *xHigherPriorityTaskWoken = pvPortMalloc(sizeof(BaseType_t));
+	if(xQueue){
+		xQueue->uxMessagesWaiting = nondet_UBaseType_t();
```
Comment (danielsn) [inline tools/cbmc/proofs/Queue/QueueGiveFromISR/QueueGiveFromISR_harness.c]: If its already unconstrained, what is this doing here?

**Comment by markrtuttle** [comment]
Closed in favor of https://github.com/aws/amazon-freertos/pull/1324.

**Pull request closed without merging**